Prod4 pod 1177

alexandrosFarhat opened the pull request:
As to avoid the nasty if statements to change npm scripts conditionally based on the platform, dependencies are used to reach a cross-platform functionality. I think the zip might be done better, but to avoid adding new environment variables, a method using npx and an installed package "bestzip" was implemented.  This kept the code neat and maintained the execFileSync call. 

Also reverted the font flip, will put it in a separate branch

JJ reviewed (changes requested):
First, thanks for addressing some of the concerns, and doing it so fast. So what's happened to #541 ? Is it no longer relevant?

At any rate, there are at least two issues here.
1. The thing with symbolic links in Windows. I would very much prefer you look for a local (as in, your own machine) solution instead of changing the code here. There are [several solutions proposed here](https://stackoverflow.com/questions/5917249/git-symlinks-in-windows), did you try them? This irrespective that copying files might or might not be a better solution than symbolic links. My point is that we should try to scope the PR correctly and do only one thing, once we've agreed on what to do.
2. Use of Linux-only commands. Two different issues here: one is using rm -rf (which I would suggest to be substituted by rimraf, if really needed) or zip (of which there are several options, one of them might be the one you've used). The point here would be to substitute these linux-only commands by equivalent cross platform dependencies.
This way two of my concerns would be addressed
1. There needs to be *very good justification* for changing the main infrastructure files such as build.js and the main package.json
2. We should try and avoid platform-dependent code paths, because in that case we would need to maintain both code paths (and test them). Using cross-platform solutions, it's up to the solution's vendor to address whatever complexity is there.

fhd commented on build/index.js:
```diff
@@ -155,9 +155,13 @@ function executeProcess(executable, args, env = process.env) {
+    const cmd = windowsEnvironment ? "npm.cmd" : "npm";
+    await executeProcess(cmd, args, { ...process.env, FORCE_COLOR: 1 });
+    const elapsed = new Date() - start;
+    logDetail(`NPM finished in ${elapsed} ms`);
```
I also noticed that this seems to be pulling in some changes that have already landed on `main` - can't entirely explain it, but it's odd.

JJ replied:
Probably a merge without rebase.

fhd commented on build/index.js:
```diff
@@ -1,5 +1,4 @@
 "use strict";
-
```
Still unrelated :)

fhd replied:
Quickly fixed it :D

fhd commented on build/index.js:
```diff
@@ -174,7 +178,6 @@ async function npmRun(script, pkg) {
 
 async function cleanPackage(pkg) {
     if (await npmRun("clean", pkg)) return;
-
```
Unrelated

fhd replied:
Quickly fixed it :D

fhd commented on build/index.js:
```diff
@@ -271,6 +274,7 @@ async function main() {
         logDetail(`🧹 ...`);
         await executeProcess("npx", ["eslint", ...eslintOptions]);
         logSuccess(command);
+
```
Unrelated

fhd replied:
Quickly fixed it :D

fhd commented on core/api/fetch-spec/package.json:
```diff
@@ -13,7 +13,7 @@
   },
   "scripts": {
-    "build": "rm -rf tsconfig.tsbuildinfo dist && tsc --emitDeclarationOnly && rollup -c",
+    "build": "shx rm -rf tsconfig.tsbuildinfo dist && tsc --emitDeclarationOnly && rollup -c",
```
Hm, I kinda like it - hadn't heard about `shx`, but at least it's only a single package as opposed to `rimraf` for removing, and who knows what for copying.

JJ replied:
Well, rimraf is a plugin replacement for rm -rf, which is precisely what we need.

JJ replied:
Also, published for the last time a year ago, version 0.3.3... [Not really my first choice for this](https://www.npmjs.com/package/shx). `rimraf` is [even older](https://www.npmjs.com/search?q=%22rm%20-rf%22), but there are other replacements. If we need to introduce some dep at the top level, I'm not sure I'd go for one with a major version 0 (or for one that's 2 years old)

fhd replied:
Hm, `shx` seems to be reasonably widely used. What I like about using `shx` instead of replacing `rm -rf` by a singularly focused package (e.g. `rimraf`) is that it allows us to move fast: We can do basic stuff with the shell _without_ having to hard code simple operations (via Rollup, a custom script, whatever). Using `shx` we wouldn't be limiting ourselves just to have a cross platform build, we could still pretty much do the same stuff we're used to, not stopping to use Rollup for a trivial copy, or which exact package to use for removing files, which for copying, ...

Reminds me of how OpenSSL is developed: AFAIK it's developed purely against OpenBSD, and then a _separate_ team adds a portability layer. Sounds weird, but I find that quite elegant actually :) Wouldn't want the cross platform build to slow us down,  that's why I find `shx` (or something like it) so appealing.

fhd commented on features/bundle/index.js:
```diff
@@ -13,8 +13,14 @@ function packageFeature({ archiveName, moduleName, artifactPath }, targetDir) {
     );
-    const args = ["-r", targetArchive, "."];
-    child_process.execFileSync("zip", args, { cwd: sourceDir });
+    const args = windowsEnvironment
```
`bestzip` does exactly that - I would argue we should _always_ use that one, not just on Windows. That invisible dependency to the `zip` binary we currently have on non-Windows systems was definitely eating at me, and now it'd be _less_ code to get rid of it - my dream :D

I suppose one _could_ argue that we could also be using something like _archiver_ directly, but I have a suspicion that'd be more code :D Given how this script is truly quite the _script_, I think _bestzip_ is a good choice actually.

JJ replied:
But then we don't need to check for Windows here, right?

fhd replied:
Yup, it's gone already in fact :)

fhd commented on features/bundle/package.json:
```diff
@@ -4,6 +4,7 @@
     "build": "node index.js"
   },
   "devDependencies": {
+    "bestzip": "^2.2.0",
```
From my understanding yes, and it uses a native ZIP binary if available, so it shouldn't change semantics on our current build servers, where we do install the ZIP binary explicitly AFAIK.

fhd replied:
Yup, confirmed that it works as expected on both macOS and our Linux-based GHA runners.

fhd commented on package.json:
```diff
@@ -11,6 +11,7 @@
     "eslint-plugin-svelte3": "^3.2.0",
     "jest": "^27.2.5",
     "prettier": "^2.3.0",
+    "shx": "^0.3.3",
```
Nice that you added it directly to the root module!

fhd commented on build/index.js:
```diff
@@ -155,9 +155,13 @@ function executeProcess(executable, args, env = process.env) {
-    executeProcess("npm", args, { ...process.env, FORCE_COLOR: 1 });
+const npm = async (...args) => {
+    const start = new Date();
+    const cmd = windowsEnvironment ? "npm.cmd" : "npm";
```
The executable just has a different name on Windows - executables traditionally have no file extension on UNIX-based systems, but they do on Windows.

Bit of a nit, but I would argue you can get rid of the `windowsEnvironment` variable and just inline the logic here, given how it's only ever used here:

```
const cmd = process.platform === "win32" ? "npm.cmd" : "npm";
```

JJ replied:
There's a simple (and local)[workaround](https://stackoverflow.com/a/13319306/891440). Again, if something can be solved locally, and we're not even supporting the platform in CI, it's probably better for the overall health of the code if it's solved that way.

JJ replied:
OK, besides the workaround (which I would really encourage you to use), now that we use `shx` at the top level, we could simply use `shelljs` and use *that* shell, which does this kind of thing for you. I still don't like to have that kind of stuff in the code. Can we do that? Please? :pray:

fhd replied:
I think that'd be great!

alexandrosFarhat replied:
Just addressed that in my final commit, let me know if that's the sort of thing you were looking for :)

fhd reviewed (changes requested):
I actually like this quite a bit! To sum up:

1. Using `bestzip` and `shx` is nice - it's a drop in replacement for the current logic, requiring very few changes. Alternatively we could use a plain ZIP library in JS code, `rimraf` as replacement for `rm -rf`, and something else as a replacement for `cp -r`, but I quite like how minimal the changes are with the two packages you chose. Windows is probably never going to be our primary development target, so I do like that we can kind of just pretend we're writing POSIX shell scripts :D
2. The symlinks had some ephemeral beauty, but I was kinda mentally prepared for us having to move over to copying resources instead, definitely less mystical, and makes the Rollup scripts readable again, too (those weird workarounds we needed...).

So yeah, I'm a fan. I _did_ notice some odd changes being added that are already on `main` -> Maybe it's time to merge `main` into this branch again? I also left a couple of comments that would still be nice to address.

But still, very nice IMHO.

JJ commented on features/demo/package.json:
```diff
@@ -2,9 +2,9 @@
+    "clean": "shx rm -rf dist",
     "prebuild": "npm run clean",
-    "build": "cp -r src dist && cp node_modules/@polypoly-eu/podjs/dist/pod.js dist"
+    "build": "shx cp -r src dist && shx cp node_modules/@polypoly-eu/podjs/dist/pod.js dist"
```
In this case maybe including a minimal rollup script would be less hacky than this.

JJ replied:
... but that could be done in another PR.

JJ reviewed (changes requested):
I have a hunch that the error in the log is related to the change in symlinks. I'd still be more comfortable if that was handled separately. I think it deserves some careful consideration, as well as handling of whatever error comes up. If @alexandrosFarhat / @alefalafel can take care of that, I'd really appreciate it. Then we can deal with whatever we use to substitute rm -rf, and whether we need specific detection of windows environment or not.

JJ commented:
Please also note that you're using two different GitHub users here, and that we usually sign the commits; there seems to be only one case where you've done so.

fhd commented:
> Please also note that you're using two different GitHub users here, and that we usually sign the commits; there seems to be only one case where you've done so.

Good point - I think this is a good candidate for squash merging anyway, so in the end we can have a single signed commit with the correct user name in it.

fhd commented:
@JJ So I think there are four changes here:

1. Copying resources at build time that used to be symlinks
2. Using `npm.cmd` and `npx.cmd` respectively on Windows
3. Introducing `shx` to make our `rm -rf`, `cp` and `cp -r` calls portable
4. Introducing `bestzip` to eliminate the invisible dependency to the `zip` executable we currently have

I think all of those things are good, but would you feel more comfortable if we split out (1)? There does seem to be a workaround for working with symlinks on Windows, as you posted, maybe we can just link to something like this from the README for now. If you're so uncomfortable with moving away from symlinks, I think this is a valid way to go. Certainly wouldn't want that to hold back (2)-(4), which are enough to make the build work on Windows, with (4) even being something I'd want to have _even if we didn't want to support building on Windows_.

alexandrosFarhat commented:
Thank you both for your constructive comments!
I reverted the changes done to substitute the symlinks and committed the change in how file zipping is done as well as removed the windows environment variable and added the npm command change as an in-line conditional as requested. 
From what we've discussed I understand  we have 3 options:
A) Just add a note for Windows users to the README that points them somewhere to figure it out themselves how to make symlinks in Git work 
B) Add a note for Windows users to the README and a script to the repo they can run to sort it out (guess it could go into build, or just at the top level) 
C) Bring back the resource copying changes
I think option B is a nice and relatively painless workaround for any new user trying to get involved with the git while maintaining symlinks. Kindly let me know what option you would recommend @JJ @fhd :)

fhd commented:
@alexandrosFarhat I prefer (C) - bringing back the resource copying changes. The more I think about it, the more it seems like the right (TM) way to do it for me. But I find it more important what @JJ prefers - he's the one dealing the most with (improving) our build system, build times and all that.

fhd commented on features/bundle/index.js:
```diff
@@ -1,9 +1,8 @@
 const fs = require("fs");
 const path = require("path");
-
+const zip = require("bestzip");
```
Did a quick code style fix here, just flipping lines 5 and 6 :)

fhd commented on build/index.js:
```diff
@@ -276,7 +277,7 @@ async function main() {
     const eslintOptions = ["--ext", ".ts,.js,.tsx,.jsx", "."];
 
-    if (!["list", "list-deps", "clean"].includes(command)) {
+    if (!["list", "list-deps"].includes(command)) {
```
This change, and the one below, were necessary for `clean` to work, now that it's using `shx`,  a root module. *I still think it's a good idea to use shx for the reasons I wrote down elsewhere*, but I guess I know someone who's not gonna like this 🙈

fhd reviewed (approved):
LGTM as far as I'm concerned!

I made some changes to get it there: 

1. Eliminated those unrelated whitespace changes I had pointed out: https://github.com/polypoly-eu/polyPod/pull/547/commits/9c82571b4c8ec393884c66380a067cf5e3536468
2. Fixed the symlinks: https://github.com/polypoly-eu/polyPod/pull/547/commits/9afee9341ecce4e0e4b1e73b1589d7449c854e1e (somehow they broke when they were brought back - like this it's identical to how they are in `main` again)
3. Fixed feature bundling (the comment says it all): https://github.com/polypoly-eu/polyPod/pull/547/commits/bb47a0b09d51b9ee2c6bc2784afb135c04219499
4. Merged `main` in and resolved the conflict: https://github.com/polypoly-eu/polyPod/pull/547/commits/8c6c5703f337a481c14801f984ec70677bab9f34
5. Fixed `./build.js clean`, which now rely on `shx`: https://github.com/polypoly-eu/polyPod/pull/547/commits/66936e8e5752104388f83a40024ac27c4ad1e17a and https://github.com/polypoly-eu/polyPod/pull/547/commits/45270764c0db07e5d194fcaa663b36aad7d77f9f

The `check_frozen_files` action failure is to be expected, given I've modified frozen files. Everything else looks good now.

JJ reviewed (approved):
Major obstacles have been cleared, I still would need to take a good look at shx, we can always work out whatever needs to be done later. Thanks a lot! (I would also appreciate you addressing my comments before or after merging, does not matter much)

fhd commented on build/index.js:
```diff
@@ -320,6 +317,10 @@ async function main() {
         return 1;
     }
+
+    if (command === "clean") {
```
I had to move it down - otherwise we would _first_ install the dependencies needed to clean some sub packages (i.e. `shx`), _then_ we would clean the root package (removing them again!), _then_ we would try to use them.

Found this issue while testing everything a bit, it basically causes all the _custom_ clean scripts some (very few, really) packages have to fail.

JJ replied:
Right. Deps are installed when build... And there's a implicit dependency for you. I think we should *really* refactor, revamp and even spin off this build script, making it an all-powerful task runner for monorepos called *baut*. It's becoming increasingly clear that we will need to modularize and refactor at some point, and that point is becoming increasingly close as we find this kind of hacks needed to make it just work.

fhd replied:
Or _baumeister_? Anyway, we can modularise this even if we just stick to _build_ for now - there's already _gyp_ for the hyper generic monorepo tool, I think we want to be specific rather than generic.

As for this hack, there's two problems: The script actually somewhat elegantly _only_ deals with packages, which includes _build_ itself. But we do _not_ treat the root package like a package, and the root dependencies aren't something the dependency tree covers either (cause it introduces dependencies even for the clean scripts).

The other problem, that @alexandrosFarhat actually found: It doesn't do anything here! This code would need to be moved _before_ the `return 0` in the try block above. I wonder why the linter didn't complain...

fhd commented on features/bundle/index.js:
```diff
@@ -13,8 +13,14 @@ function packageFeature({ archiveName, moduleName, artifactPath }, targetDir) {
+        destination: targetArchive,
+        cwd: sourceDir,
+    }).catch((error) => {
+        console.error(error.stack);
```
I actually wondered the same :D In short: Yes, it gives a slightly nicer error message, otherwise Node just talks about there being an "uncaught promise failure", this way it's a wee bit nicer, and clear enough where the error is from. The most elegant thing would probably be to turn this into an async function, and catch any exceptions thrown by it below, doing some proper error logging for those cases. But I didn't want to overengineer this little script too much at this time :D

JJ commented:
It's failing now on all pistons. I'll try to bring it back to green soonish

JJ commented:
I guess @alexandrosFarhat doesn't have privs for merging it now; the last commit is mine so I guess the ball is on your field for merging, @fhd

fhd commented on build/index.js:
```diff
@@ -159,11 +160,7 @@ function executeProcess(executable, args, env = process.env) {
-        ["--no-update-notifier", "--no-fund", ...args],
-        { ...process.env, FORCE_COLOR: 1 }
-    );
+    await executeProcess("npm", ["--no-update-notifier", "--no-fund", ...args]);
```
@JJ Did you mean to remove the glorious `FORCE_COLOR`? :D

JJ replied:
Not me! Put it back!

fhd reviewed (approved):
Fixed that last thing - LGTM now. One last question for @JJ, then I'll happily merge :D

JJ commented:
:clap: to @alexandrosFarhat for their first merge! :1st_place_medal:
